### PR TITLE
Add support for dynamic themes

### PR DESCRIPTION
**Pull request by artlowel**
## References
* Handles #729, but doesn't completely solve it yet.

## Description
This PR adds support for themes. And the ability to configure which theme is active based on the route

## Instructions for Reviewers
This PR makes the home page, home news section (since I already made it themeable for my proof of concept, I kept it) and item page themeable. It also moves everything that can be affected by a theme from AppComponent to a new [RootComponent](https://github.com/atmire/dspace-angular/blob/dynamic-theme-support/src/app/root/root.component.ts), and makes that themeable as well.

It adds a [template theme in the themes folder called "custom"](https://github.com/atmire/dspace-angular/tree/dynamic-theme-support/src/themes/custom) that has an implementation for each of the themeable components. By default the style and template properties for those themed components refer to the unthemed versions of those components, so you won't see a difference. But each also has its own html and scss file you can swap to by uncommenting them. 

You can add global styles and override variables for the theme in [the /styles folder](https://github.com/atmire/dspace-angular/tree/dynamic-theme-support/src/themes/custom/styles).

You can configure theme rules in the environment file based on regex, uuid or handle. [environment.common.ts contains an example in comments](https://github.com/atmire/dspace-angular/blob/dynamic-theme-support/src/environments/environment.common.ts#L231-L253) for each of these.

You can test this PR by enabling some theme rules, and modifying the custom folder directly or duplicating the custom folder and renaming it for your own theme. If you do the latter, make sure to also add the new theme to [the styles array in angular.json](https://github.com/atmire/dspace-angular/blob/dynamic-theme-support/angular.json#L49-L60)

### Creating a theme
Based on [@tdonohue's comment below](https://github.com/DSpace/dspace-angular/pull/1035#pullrequestreview-600989619), here's a first pass at a guide to theming using this PR:

1. Either start with `src/themes/custom` or create a new `src/themes/[mytheme]` directory, by copying the `src/themes/custom` theme and renaming it
     * If you do the latter, edit `angular.json` to add your new theme to the `styles` array.  Make sure the `bundleName` in that `styles` array is of the format: `${themeName}-theme`, e.g. `custom-theme` for a theme named `custom`.
2. Enable your theme in `environment.*.ts`, by adding it to the `themes` array, and optionally defining a handle/regex/uuid to limit its context
3. To make global CSS changes, modify your theme's `styles` folder
4. To include new fonts, images or i18n in your theme put them in the assets subdirectory of your theme. e.g. if you create `src/themes/custom/assets/images/logo.png` it will be deployed under `/assets/custom/images/logo.png`
5. To theme a specific component, find its `*.ts` file in your theme folder. By default these template themed components still use the scss and html files for the original component, to allow you to choose which you want to override
6. If you want to override the style in your themed component, uncomment the commented `styleUrls` line, and comment the active one. From that point on the `*.scss` file for that component in the theme folder will be used.
7. If you want to override the html in your themed component, uncomment the commented `templateUrl`, and comment the active one. From that point on the `*.html` file for that component in the theme folder will be used. 
8. Rebuild (if you ran `yarn start`) or wait for app to rebuild automatically (if you ran `yarn start:dev`) & try it out.
9. Any components that you _don't_ want to override in your theme can be removed from your theme folder after you're ready. Leaving them doesn't do much harm, but removing them will slightly reduce the build time and bundle sizes. Also remove the references to the removed components from `theme.module.ts` and `entry.components.ts` in your theme

#### 2 types of style variables
In order for unthemed dspace components to respond to changes in the global theme CSS I had to switch from using Sass variables (which are evaluated during the build) to CSS variables (which are evaluated on the fly). That isn't a major problem in and of itself, it is likely the future. However bootstrap doesn't use CSS variables, it uses Sass. So in order to override bootstrap variables we _still_ have to use Sass. So the custom theme has a separate file for sass variable overrides and for css variable overrides (DSpace's own variables). It also means we have to have a mapping file to map all bootstrap sass variables on to CSS variables. 

#### Unused warning
Every theme module will generate the following warning, during build, and in the browser when in dev mode:
```
WARNING in src/themes/custom/theme.module.ts is part of the TypeScript compilation but it's unused.
Add only entry points to the 'files' or 'include' properties in your tsconfig.
``` 
It is exactly the point that these modules are unused. Their purpose is to have a context for our themed components to be compiled in, but they aren't needed to run the app: the components are lazy-loaded directly, without the module. A way to get rid of this problem would be to create a custom project in angular.json for each theme, but that would make it a lot more complicated to add an additional theme, as well as make the builds take longer. Alternatively, conditionally lazily loading each theme module somewhere in the app, under a condition that never occurs would likely also solve the issue, but seems dirty, and would also require us to reference each custom theme separately in another place. So for now I kept the warning.

## Checklist
- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Comment by tdonohue** [comment]
@artlowel : Just a quick note that this has a small merge conflict.

**Comment by artlowel** [comment]
The tests wouldn't work due to [this incident](https://www.githubstatus.com/incidents/xn0sd2x4nd7f). I tried to switch the base to something else and back, but that didn't fix it (in fact it broke LGTM, as it picked up one switch but not the other). A small new commit seems to do the trick

**Review by tdonohue — CHANGES_REQUESTED** [review]
@artlowel : Overall, this honestly looks good.  I have glanced at the code and it looks reasonable. I ran some tests too (all using `yarn start`) and it seems to mostly work.  I did find a few very small things though:

* Unless I'm mistaken (or using it wrong), the `src/themes/custom/styles/_theme_sass_variable_overrides.scss` doesn't appear to work.  I uncommented the examples in that file & did not seem them override the defaults.
    * In contrast, the `_theme_css_variable_overrides_scss` file **does** work.  So, for example, I added `--blue: #210af1;` into the `:root` section and saw that it overrode the blue color.
2. While I appreciate the example overridden components in testing this PR, I actually wonder whether they should be removed from this PR.  In my testing, several times I stumbled over the fact that I had to remember to *edit* the overridden `*.component.ts` to change its `styleUrls` or `templateUrl` appropriately.  I definitely understand why you wanted to put these examples here, just noting they are clunky to work with as a "user"....and therefore I question whether they are helpful in the long term. (I could be missing something here though)

As I said, overall this is working.  I think this PR could be improved though by better documenting the steps to create a theme (because I'm not even sure if I understand it fully yet). Here's what I *think* is how you are supposed to create a theme:

1. Either start with `src/themes/custom` or create a new `src/themes/[mytheme]` directory, copying over the `src/themes/custom` theme.
     * If you do the latter, you MUST edit `angular.json` to add your new theme to the `styles` array.
     * It you do the latter, it seems like you ideally should copy the 'custom' theme directory to a new location, as it has some important files you MUST have to make theming work, namely the `theme.module.ts`
2. Enable your theme in `environment.*.ts`, by adding it to the `themes` array (and optionally defining a handle/regex/uuid to limit its context.  I tested the UUID option and it works well..haven't tried the other two yet)
3. At a minimum, you likely want to modify your theme's `styles` folder to make CSS changes.
4. It appears (I could be wrong) you cannot include new fonts, images or i18n in your theme directly...instead those need to all go into `src/assets/` in order to be used by your theme.
5. To override a specific component, create a folder of the same name/path under `src/themes/[mytheme]/app/` as the original component under `src/app/`  (Obviously a `themed-*` version of that component must exist for this to work)
6. Then copy overridden component files into that directory. YOU MUST COPY OVER THE *.component.ts, and likely want the *.html and *.scss. But, it seems like you should *skip* copying over other files, especially the `themed-[component].component.ts`. 
7. It also appears this overridden component MUST be listed in the `DECLARATIONS` of `theme.module.ts` in your theme, or else it won't be used.  (However, I assume we'll do our best to keep this theme.module.ts up-to-date for users, so most can skip this step)
8. Rebuild (if you ran `yarn start`) or wait for app to rebuild automatically (if you ran `yarn start:dev`) & try it out.

I think it'd be good to start documenting the steps here, as I don't think it's entirely straightforward yet (at least not to me).  Step 5-6 above is the most confusing part, as it seems like you don't necessarily want to copy the entire component folder over into your theme. But, you *MUST* make sure you minimally have the main `*.component.ts` copied over.  I'm not sure there's any way to make this easier, but clearly documenting the steps might be the best place to start.

Overall, good work here. I think this is looking good, we just need to look at making the instructions much clearer (maybe we should draft them in this PR's description?)

**Comment by artlowel** [comment]
>Unless I'm mistaken (or using it wrong), the src/themes/custom/styles/_theme_sass_variable_overrides.scss doesn't appear to work. I uncommented the examples in that file & did not seem them override the defaults.

That's strange, for me just uncommenting the `$blue` line and rebuilding in prod mode does have an effect. It doesn't affect the link color as that is based on `$info` in dspace iirc. But it does affect button colors, and pagination etc:
image

>While I appreciate the example overridden components in testing this PR, I actually wonder whether they should be removed from this PR. In my testing, several times I stumbled over the fact that I had to remember to edit the overridden *.component.ts to change its styleUrls or templateUrl appropriately. I definitely understand why you wanted to put these examples here, just noting they are clunky to work with as a "user"....and therefore I question whether they are helpful in the long term. (I could be missing something here though)

Until we have our [own schematic for generating themed components using cli](https://medium.com/@tomastrajan/total-guide-to-custom-angular-schematics-5c50cf90cdb4), I think it would be more error prone without them. I consider the custom theme to be a template where every themable component has an implementation. If you want to create a new theme either work on that template directly, or use it as a place to grab a starting point for your own versions of these components. With it, all you need to do to start customizing a component is swap the comments on styleUrls and templateUrl around. Without it, you'd have to do the import `as BaseComponent` by hand,  figure out the exact amount of `../`-es you need to reference the base template or scss if you only want to override one of the two, remember to add it to theme.module or entryComponents etc. I've asked our devs that are working on part 2 of this task to generate similar example components for every themeable component in dspace in that custom folder.

>It appears (I could be wrong) you cannot include new fonts, images or i18n in your theme directly...instead those need to all go into src/assets/ in order to be used by your theme.

You're right, I forgot about assets. I'll see if I can't support putting those in the themes folders as well.

**Comment by artlowel** [comment]
@tdonohue I added support for assets, and started from your comment to add a guide to theming to the PR description

**Review by tdonohue — CHANGES_REQUESTED** [review]
@artlowel : After your most recent changes to support images/fonts, I'm no longer able to build this PR using either `yarn start` or `yarn start:dev`.  The full error stack looks like this:
```
[error] ValidationError: CopyPlugin Invalid Options

options should be array

    at validateOptions (D:\programming\dspace-angular\node_modules\copy-webpack-plugin\node_modules\schema-utils\src\validateOptions.js:32:11)
    at new CopyPlugin (D:\programming\dspace-angular\node_modules\copy-webpack-plugin\dist\index.js:26:30)
    at Object.<anonymous> (D:\programming\dspace-angular\webpack\webpack.common.ts:60:5)
    ....
 ```

So, it appears to me like we have an accidental bug in our `webpack.common.ts`.  

I did want to verify though that your prior responses all make sense...now that we had the deeper discussion about theming, I see my original "assumptions" were a bit flawed, and I better understand the direction here (and it seems like a good direction).  Once the build is fixed, I'll retest building a new theme based on the instructions in this PR's description.

**Comment by artlowel** [comment]
@tdonohue I had to update the CopyPlugin to support wildcards. Is it possible you forgot to re-run `yarn clean` and `yarn install`?

File: src/modules/app/browser-app.module.ts [hunk 1]
```diff
@@ -51,12 +50,12 @@ export function getRequest(transferState: TransferState): any {
     HttpClientModule,
     // forRoot ensures the providers are only created once
     IdlePreloadModule.forRoot(),
-    RouterModule.forRoot([], {
-      // enableTracing: true,
-      useHash: false,
-      scrollPositionRestoration: 'enabled',
-      preloadingStrategy: NoPreloading
-    }),
+    // RouterModule.forRoot([], {
```
Comment (atarix83) [inline src/modules/app/browser-app.module.ts:53]: why is this section commented?

File: src/modules/app/server-app.module.ts [hunk 2]
```diff
@@ -42,9 +41,9 @@ export function createTranslateLoader() {
     BrowserModule.withServerTransition({
       appId: 'dspace-angular'
     }),
-    RouterModule.forRoot([], {
-      useHash: false
-    }),
+    // RouterModule.forRoot([], {
+    //   useHash: false
```
Comment (atarix83) [inline src/modules/app/server-app.module.ts:45]: the same as above ?

**Review by atarix83 — CHANGES_REQUESTED** [review]
thhank @artlowel for this work.

the code looks good to me, I've only few inline comments.

I tried to use the costum theme by overriding some components without success. Here the steps that I followed : 
- I've enabled costum theme by uncommentig [these lines](https://github.com/atmire/dspace-angular/blob/1bb2d5da294312a398b82badb0115aef800c1098/src/environments/environment.common.ts#L251-L253) and commetting [these lines](https://github.com/atmire/dspace-angular/blob/1bb2d5da294312a398b82badb0115aef800c1098/src/environments/environment.common.ts#L255-L258) in my environment file
- I've changed the https://github.com/atmire/dspace-angular/blob/dynamic-theme-support/src/themes/custom/app/%2Bhome-page/home-page.component.ts in order to use the custom template instead of the base one
- I've changed the https://github.com/atmire/dspace-angular/blob/dynamic-theme-support/src/themes/custom/app/%2Bhome-page/home-page.component.html file

After those steps I'd expect to see the changes made instead I continue to see the original base template. 
@artlowel Did I make a mistake or is there something wrong?

File: src/app/+home-page/home-page.component.ts [hunk 3]
```diff
@@ -10,7 +10,7 @@ import { Site } from '../core/shared/site.model';
   templateUrl: './home-page.component.html'
 })
 export class HomePageComponent implements OnInit {
-
+  testInput = 'Bingo!';
```
[the remark this answers is not in the document]
Comment (artlowel) [reply src/app/+home-page/home-page.component.ts]: You're right. Removed

File: src/app/+home-page/home-page.component.ts [hunk 4]
```diff
@@ -23,4 +23,8 @@ export class HomePageComponent implements OnInit {
       map((data) => data.site as Site),
     );
   }
+
+  onOutput(event: string) {
```
[the remark this answers is not in the document]
Comment (artlowel) [reply src/app/+home-page/home-page.component.ts]: Removed

File: src/modules/app/browser-app.module.ts (hunk 1 above)
Comment (artlowel) [reply src/modules/app/browser-app.module.ts]: Also to test something. I've reverted it. Thanks for catching that!

File: src/modules/app/server-app.module.ts (hunk 2 above)
Comment (artlowel) [reply src/modules/app/server-app.module.ts]: Same as above 🙂

**Comment by artlowel** [comment]
@atarix83 If I try the steps you described, I see the custom home page. This is what I did: https://github.com/atmire/dspace-angular/commit/3d1c401aac8d58a069b00c4b2f151512d4056be0

Perhaps you still have a themes section in an environment.dev.ts or environment.prod.ts file that overrides the version in environment.common.ts ?

**Comment by atarix83** [comment]
@artlowel if I do as your example it works, but instead of changing the theme in the `environment.common.ts` if I do it in the `environment.dev.ts` or `environment.prod.ts` it doesn't work. Is it an issue or is how it is supposed to work?

**Review by atarix83 — COMMENTED** [review]
@artlowel I found another issue on navbar menu with small screen view. Probably is due to the new css variable, here a screenshot of the problem
Schermata da 2021-03-08 17-09-11

**Comment by artlowel** [comment]
>@artlowel if I do as your example it works, but instead of changing the theme in the environment.common.ts if I do it in the environment.dev.ts or environment.prod.ts it doesn't work. Is it an issue or is how it is supposed to work?

Perhaps you're running into this issue? https://github.com/DSpace/dspace-angular/issues/909
Check the generated environment.ts file to see what's in the themes array. 

>@artlowel I found another issue on navbar menu with small screen view. Probably is due to the new css variable

It turns out you [can't use CSS variables in media queries](https://stackoverflow.com/questions/40722882/css-native-variables-not-working-in-media-queries). That isn't supported by the spec. So I had to replace them with sass variables again. 

This doesn't mean you can't customize the grid breakpoints _at all_. However it does mean you can't have different themes on the same DSpace instance with different grid breakpoints. You can set them in `src/styles/_bootstrap_variables.scss`, but not vary them per theme.

**Comment by tdonohue** [comment]
@artlowel : I'm seeing the same as @atarix83 .  I can change the theme if I do so in `environment.common.ts`.  But, even a simple `environment.dev.ts` like this doesn't work (I've added obvious HTML changes to 'custom' and verified they do not load in `environment.dev.ts`, but do load in `environment.common.ts`):
```
export const environment = {
  themes: [
    {
       name: 'custom'
    },
    {
      name: BASE_THEME_NAME
    },
  ],
};
```

Also, a simple question (perhaps worth documenting): It looks like the `bundleName` assigned to new themes in `angular.json` doesn't seem to matter?  I had expected that it must match the name of the theme, but I see you have [`bundleName: custom-theme`](https://github.com/DSpace/dspace-angular/pull/1035/files#diff-2d2675bb4687601a5c7ccf707455132f90f3516a33716185687e5c41df59ac7dR58), while the theme itself is named `custom`.  Is there a required naming convention for this `bundleName`?

You were correct that my issues on Friday were because I forgot to run `yarn clean` prior to running `yarn install`.  Today, everything is looking much better. There are still some small bugs, but the general direction is good.

**Comment by artlowel** [comment]
>It looks like the bundleName assigned to new themes in angular.json doesn't seem to matter? I had expected that it must match the name of the theme, but I see you have bundleName: custom-theme, while the theme itself is named custom. Is there a required naming convention for this bundleName?

Yes, the bundleName is the name of the css file that webpack will generate. it has to be `${themeName}-theme` because that's what app.component will assume it is [when swapping out css files](https://github.com/atmire/dspace-angular/blob/dynamic-theme-support/src/app/app.component.ts#L198). I opted to add the `-theme` to lower the risk of clashes with other, pre-existing css files. E.g. there's a chance that `base.css` already exists for some other purpose.

**Review by tdonohue — COMMENTED** [review]
@artlowel : I did some more testing today.  Overall, this is working well, but I've still found some issues:

1. The `_theme_sass_variable_overrides.scss` file still doesn't seem to work for me. I tried a more extreme change by setting the primary/secondary colors to red & green and used `yarn start:dev`.  Simple color changes also are not seemingly being picked up I'm still seeing no overrides:
    ```
    $theme-colors: (
        primary: $red,
        secondary: $green,
        success: $green,
        info: $cyan,
        warning: $yellow,
        danger: $red,
        light: $gray-100,
        dark: $dark
    ) !default;
    ```
    I also attempted a font change by adding this to that same file, and didn't see a font change.  (I borrowed these from the old mantis theme):
    ```
    @import url('https://fonts.googleapis.com/css?family=Source+Sans+Pro:200,200i,300,300i,400,400i,600,600i,700,700i,900,900i&subset=cyrillic,cyrillic-ext,greek,greek-ext,latin-ext,vietnamese');

    $font-family-sans-serif: 'Source Sans Pro', -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
    ```

2. In "good news", the component overlays are working great.  I've noticed that we also can import the default SCSS into the theme's SCSS.  So, this line works in `custom/app/+home-page/home-page.component.scss`: 
    ```
    @import '../../../../app/+home-page/home-page.component.scss';
    ```
    This import is nice because it means we could simply set `styleUrls: ['./home-page.component.scss'],` **by default** to let anyone override/modify the SCSS immediately.  They'd still need to uncomment & manually copy over the HTML template though if that needed modification.

**Comment by artlowel** [comment]
@tdonohue I tried your changes in 1. in your previous comment. And the font just works for me. Overriding the $theme-colors as you did there however is invalid sass. In sass you can't override a variable once it's set. So in order for our theme's variables to win out over bootstrap's defaults, we have to [import _theme_variables_overrides.scss first in theme.scss](https://github.com/atmire/dspace-angular/blob/dynamic-theme-support/src/themes/custom/styles/theme.scss#L3-L4)

As a consequence, $red, $green, etc, aren't yet defined there, unless you define them. But when you do, it works. See my test here, that takes your $theme-colors map, but defines those colors above it: https://github.com/atmire/dspace-angular/commit/d212f7523e5d2d92521a89c4f8e27dbfe38faa94

**Review by atarix83 — APPROVED** [review]
@thanks for addressing my feedback

I can confirm the issue with navbar is resolved

> Perhaps you're running into this issue? #909 Check the generated environment.ts file to see what's in the themes array.

yes this was the issue in the environment.ts generated there were both base and custom theme enabled. 

I've also tested the possibilities to enabled custom theme for a specific uuid and it works properly

**Review by tdonohue — APPROVED** [review]
👍 @artlowel : Overall, I approve of this PR & I don't wish to hold it up as it mostly works great.  

However, I should note that I still cannot get `_theme_sass_variable_overrides.scss` to work for me.  I suspect I must be doing something wrong, or perhaps something is undocumented here.  I've tried the following and have not seen any changes in the UI theme at all:

1. First, I've run `yarn clean`, `yarn install`
2. Enabled the `custom` theme just by uncommenting this in my `environment.common.ts` themes array: 
    ```
    {
    name: 'custom'
    },
    ```
3. Run `yarn start:dev` to start in dev mode.  Everything shows with the default theme.
4. Copy your sample changes from https://github.com/atmire/dspace-angular/commit/d212f7523e5d2d92521a89c4f8e27dbfe38faa94 into  `src/themes/custom/styles/_theme_sass_variable_overrides.scss` file
5. Yarn notices the changes and recompiles.  But, I see no changes in my web browser (Chrome).  I do a hard refresh (Ctrl + reload) and also click around to different pages to ensure there's no caching...but it still all looks like the default theme.
6. Tried a more noticeable font (Stick: https://fonts.google.com/specimen/Stick) just to see if I was overlooking the font change.  Changed the font lines in  `_theme_sass_variable_overrides.scss` to be: 
    ```
    @import url('https://fonts.googleapis.com/css2?family=Stick&display=swap');
    $font-family-sans-serif: 'Stick', sans-serif;
    ```
7. Again, Yarn saw the changes and rebuilt.  But, again, the new font is not showing up in my browser.
8. I even tried stopping yarn entirely and rerunning `yarn start:dev`, but changes are still not shown in the browser.

I feel like I must be either doing something wrong, or have something different in my environment which is causing a different behavior.  Nonetheless, I'm OK with this PR overall...and this is the _only issue_ I'm hitting. So, I'm fine to merge this and attempt to debug as a follow-up PR.

**Comment by tdonohue** [comment]
Because other PRs build on this, I'm going to merge this **as is**.  My [prior feedback](https://github.com/DSpace/dspace-angular/pull/1035#pullrequestreview-609078634) on a possible bug (or clarification of documentation) could either be moved into the follow-up PR #1042 or logged as a separate ticket (@artlowel : I'd appreciate your advice on next steps assuming you are able to find a way to reproduce the issue I'm seeing)

**Pull request merged**